quic/s2n-quic-core/src/recovery/congestion_controller.rs

```diff
@@ -36,6 +39,57 @@ impl<'a> PathInfo<'a> {
     }
 }
 
+/// Wrapper around a `ConnectionPublisher` that forwards congestion control related
+/// events to the inner publisher with the necessary context.
+pub struct Publisher<'a, Pub: event::ConnectionPublisher> {
```

Comment:
One thing you could do in order to get rid of the need to pass around both publishers is make the congestion controller Publisher a trait and then have a PathPublisher that implements the trait as you've got here. I don't think it's blocking but it might be a bit cleaner IMO. Something like this:

```rust
pub trait Publisher {
    fn on_slow_start_exited(&mut self, cause: SlowStartExitCause, congestion_window: u32);
    ...
}

pub struct PathPublisher<'a, Pub: event::ConnectionPublisher> {
    publisher: &'a mut Pub,
    path_id: path::Id,
}

impl<'a, Pub> Publisher for PathPublisher<'a, Pub> {
     ...
}

pub trait CongestionController {
    fn on_rtt_update<Pub: Publisher>(
        &mut self,
        time_sent: Timestamp,
        now: Timestamp,
        rtt_estimator: &RttEstimator,
        publisher: &mut Pub,
    );
    ...
}
```

Reply:
I like that better, let me try, thanks!